tests/codex/testvalidation.nim

```diff
@@ -41,12 +56,76 @@ asyncchecksuite "validation":
   test "the list of slots that it's monitoring is empty initially":
     check validation.slots.len == 0
 
+  for (validationGroups, groupIndex) in [(100, 100'u16), (100, 101'u16)]:
+    test "initializing ValidationConfig fails when groupIndex is " &
+        "greater than or equal to validationGroups " &
+        fmt"(testing for {groupIndex = }, {validationGroups = })":
+      let groups = ValidationGroups(validationGroups).some
+      let validationConfig = ValidationConfig.init(
+          maxSlots, groups = groups, groupIndex = groupIndex)
+      check validationConfig.isFailure == true
+      check validationConfig.error.msg == "The value of the group index " &
+          "must be less than validation groups! " &
+          fmt"(got: {groupIndex = }, groups = {!groups})"
+  
+  test "initializing ValidationConfig fails when maxSlots is negative":
+    let maxSlots = -1
+    let validationConfig = ValidationConfig.init(
+        maxSlots = maxSlots, groups = ValidationGroups.none)
+    check validationConfig.isFailure == true
+    check validationConfig.error.msg == "The value of maxSlots must " &
+        fmt"be greater than or equal to 0! (got: {maxSlots})"
+  
+  test "initializing ValidationConfig fails when maxSlots is negative " &
+      "(validationGroups set)":
+    let maxSlots = -1
+    let validationConfig = ValidationConfig.init(
+        maxSlots = maxSlots, groups = validationGroups, groupIndex)
+    check validationConfig.isFailure == true
+    check validationConfig.error.msg == "The value of maxSlots must " &
+        fmt"be greater than or equal to 0! (got: {maxSlots})"
+
+  test "group index is irrelevant if validation groups are not set":
+    randomize()
+    let groupIndex = rand(uint16.high.int).uint16
```

Comment:
Please, do not use random in unit tests. If you want to do fuzz testing, then please create a new category of testing (after "unit" and "integration" tests) and have a proper setup that would ensure reproducibility of the tests.

From my end, "random values" in tests make you sloppy because you will not think about the edge case values that should be tested.

For example, a writeup on why not to use random numbers in tests: https://scotthannen.org/blog/2024/05/20/dont-use-random-numbers-in-tests.html

Reply:
Thanks for a nice article. It makes lots of sense in advising to focus on clear boundaries. Our case here is different though. There are no boundaries except for the limits of `uint16` type. The test should never fail, regardless of the value provided. You cannot really check boundaries, because there are no boundaries to check - value `0` value `65535` and value `1265` all have equal weight. No single value in the whole range of `uint16` is special. The alternative to not using random here would be to check the whole range of `uint16`, which I feel is an overkill. Thus, in my opinion, to communicate the intended behaviour, a random value feels like a good compromise. Even when there are boundaries, in some cases you could do both (checking boundaries and using random number) when the test is simple, the random value is reported and easy to interpret. Using random in a unit test is not by definition bad... "avoid" is what I would probably say.

So, I am not sure how to make it better... Maybe we should just discuss this shortly.

Reply:
Maybe this: let's make `groupIndex` an optional. In this case, when `validationGroups` is not set **and** `groupIndex` is set (i.e. it has value different than `None`) then construction of the `ValidationConfig` should fail. This has stricter semantics can be tested without using any random things. What do you think?

Reply:
After meeting with Adam, we decided to settle down on some sort of a compromise: we remove this test, and the long complex one, while keeping the one discussed [here](https://github.com/codex-storage/nim-codex/pull/890#discussion_r1784291403).